### PR TITLE
Fixed #296 Converter support for PyObject

### LINKED ISSUE #296: Python.Runtime.Converter support for converting from/to PyObject
I have a C# class, lets call it MyCSClass, that I intend to be inherited from a Python class and used like this:
```py
class MyPythonClass(MyCSClass):
   def __init__(self):
      self.MyAttribute = 5
      self.Callback(self)
```
When an instance of MyPythonClass gets passed to C#, it gets converted to a MyCSClass and is stripped of all the attributes. Even if the method called in CS takes it as a dynamic or PyObject. An example of this could be the __init__ method above calling into the following C# code:
```cs
public class MyCSClass {
   protected void Callback(dynamic self){
       Console.WriteLine("A dynamic attribute: {0}", self.MyAttribute);
   }
}
```

However, the instance assigned to 'self' is an MyCSClass and not a dynamic MyPythonClass. Replace ```dynamic``` with ```PyObject``` and you get an exception. If suggest the the converter should support PyObject, so that callbacks like these are made possible.

Then I would have to do this instead, which is OK for me:
```cs
public class MyCSClass {
   protected void Callback(PyObject self){
       Console.WriteLine("A dynamic attribute: {0}", ((dynamic)self).MyAttribute);
   }
}
```

Python Version: ```Python 2.7.11 (v2.7.11:6d1b6a68f775, Dec  5 2015, 20:32:19) [MSC v.1500 32 bit (Intel)] on win32```
OS: Windows 7 Enterprise

### PR DESCRIPTION
**Pull request by rmadsen-ks**
Work in progress.

Note, testing has not been done, except for exploratory style. How is testing supposed to be done?

### CODE REVIEW

**Comment by denfromufa** [comment]
@rmadsen-ks you can look at callbacktest.cs and test_callback.py for testing this.

**Comment by denfromufa** [comment]
@rmadsen-ks do you agree to transition from Zope to MIT license?

**Comment by rmadsen-ks** [comment]
@denfromufa, Test passed. I have added a test verifying that the fix does what it should. 

I agree to transition to MIT license.

File: src/tests/test_class.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 import System
 import six
 from Python.Test import ClassTest
-from System.Collections import Hashtable
+from System.Collections import Hashtable
```
Comment (denfromufa) [inline src/tests/test_class.py]: where is this needed?

Comment (rmadsen-ks) [inline src/tests/test_class.py]: It was not needed by me, but it seems I had added a windows line-ending. Its fixed now.

**Comment by denfromufa** [comment]
how about garbage collection, is CLR or Python responsible? If Python, then I do not see Decref, so it would leak memory?

**Comment by rmadsen-ks** [comment]
I have tested callbacks rapidly and the ref count does not increase permanently. Neither does it seem to leak when many new objects are made. I have tried creating 50M objects with no noticable leak.

 I should not need to Decref for the following reason:

As the comment in pyobject.cs says, the PyObject assumes ownership. So if we want to convert a python object to a PyObject using ToManagedValue, we must increase the refcount, because we are giving it to the constructor of PyObject. So I implemented that part. The PyObject decreases the ref count when the garbage collector cleans it up, which happens eventually. So when a new C# PyObject has been created from a python object, it will eventually be cleaned up by the .NET GC. Is some cases ```using(){}``` can also be used, to give more predictable memory performance. 

In the same way, if we want to convert C# PyObject to a python object, we increase the ref count, because now Python has an extra reference to it, so it needs ownership. Python will clean that up on its own when it goes out of scope.

**Comment by denfromufa** [comment]
@tonyroberts @vmuriart @filmor please review!

File: src/testing/callbacktest.cs [hunk 2]
```diff
@@ -27,5 +27,19 @@ public string Call_simpleDefaultArg_WithEmptyArgs(string moduleName)
                 return module.simpleDefaultArg();
             }
         }
+    }
+
+    //==========================================================================
+    // Tests calling from Python into C# and back into Python using a PyObject.
+    // SelfCallbackTest should be inherited from a Python class.
```
Comment (rmadsen-ks) [inline src/testing/callbacktest.cs]: Comment has been changed.

**Comment by denfromufa** [comment]
@rmadsen-ks this pull request does NOT resolve the issue described in #94:

```
dta@dta-Inspiron-N5050 ~/Downloads $ sudo pip install pythonnet-fix-develop-296-convert-PyObject.zip -U --force
Processing ./pythonnet-fix-develop-296-convert-PyObject.zip
Installing collected packages: pythonnet
  Found existing installation: pythonnet 2.2.0
    Uninstalling pythonnet-2.2.0:
      Successfully uninstalled pythonnet-2.2.0
  Running setup.py install for pythonnet ... done
Successfully installed pythonnet-2.2.0
You are using pip version 8.1.1, however version 9.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
dta@dta-Inspiron-N5050 ~/Downloads $ python
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import clr
>>> from System import Object, Array
>>> import sys
>>> a = Array[Object]([sys]) ; print a[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: value cannot be converted to Object
```

**Comment by rmadsen-ks** [comment]
Hi, I'm not suggesting that it does. You want #94 fixed as part of this?

**Comment by rmadsen-ks** [comment]
@denfromufa, is there anything I need to do to move this further or do I need to find another solution?

**Comment by denfromufa** [comment]
@rmadsen-ks your pull request is completely valid.

**Comment by filmor** [comment]
To me this also looks fine, but why do you introduce a new `RefCount` property? Either remove it or add tests for it, after that we can merge.

**Comment by rmadsen-ks** [comment]
@filmor, I found it useful for verifying that I was not leaking reference counts. Without it, it is hard to get out the reference cound from C# as the NativeMethods are internal. It is not critical for me, so we can remove it.

**Comment by denfromufa** [comment]
@rmadsen-ks can you please open a new PR? I made a mistake and merged this pull request without seeing @filmor concern.

**Comment by rmadsen-ks** [comment]
It is here #306

**Pull request merged**